backend-rust/src/transaction_reject.rs

```diff
@@ -525,322 +544,432 @@ pub struct PoolClosed {
                         dummy: true,
                     },
-                ))
+                )
             }
-            RejectReason::PoolClosed => Ok(TransactionRejectReason::PoolClosed(PoolClosed {
+            RejectReason::PoolClosed => TransactionRejectReason::PoolClosed(PoolClosed {
                 dummy: true,
-            })),
+            }),
+        };
+        let value = serde_json::to_value(&reason)?;
+        Ok(Self::Ready(value))
+    }
+
+    pub async fn process(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<serde_json::Value> {
+        let reason = match self {
+            Self::Ready(reason) => reason.clone(),
+            Self::RejectedReceive(prepared) => {
+                let reject = TransactionRejectReason::RejectedReceive(prepared.process(tx).await?);
+                serde_json::to_value(reject)?
+            }
+        };
+        Ok(reason)
+    }
+}
+
+/// Reject receive which require processing the contract update message
+/// using the smart contract module schema before insertion.
+pub struct PreparedRejectedReceive {
+    reject_reason:    i32,
+    contract_address: ContractAddress,
+    receive_name:     String,
+    message_as_hex:   String,
+}
+
+impl PreparedRejectedReceive {
+    pub async fn process(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<RejectedReceive> {
+        // Handle and store errors
+        let (message, message_parsing_error) = match self.process_message(tx).await {
+            Ok(message) => (message, None),
+            Err(error) => (None, Some(error.to_string())),
+        };
+
+        Ok(RejectedReceive {
+            reject_reason: self.reject_reason,
+            contract_address: self.contract_address,
+            receive_name: self.receive_name.clone(),
+            message_as_hex: self.message_as_hex.clone(),
+            message,
+            message_parsing_error,
+        })
+    }
+
+    /// Parse the message using the smart contract module of the smart contract
+    /// instance.
+    async fn process_message(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<Option<String>> {
+        if self.message_as_hex.is_empty() {
+            return Ok(None);
         }
+        let schema = sqlx::query_scalar!(
+            "SELECT
+                schema
+            FROM contracts
+                JOIN smart_contract_modules
+                    ON smart_contract_modules.module_reference = contracts.module_reference
+            WHERE index = $1 AND sub_index = $2",
+            i64::try_from(self.contract_address.index.0)?,
+            i64::try_from(self.contract_address.sub_index.0)?
+        )
+        .fetch_one(tx.as_mut())
+        .await?;
+        let Some(schema) = schema else {
+            // No schema found in the smart contract module.
+            return Ok(None);
```

Comment:
For a smart contract developer, it can be helpful for debugging if the parsing is not successful due to if no `schema` is in the contract module or `NoContractInModule`, `NoReceiveInContract`, or `NoParamsInReceive` error occurred (these cases are all logged as `None`). Should we consider logging them as `message_parsing_error`.

Reply:
I have changed the error into a status message instead.
Please have a look